### DISCUSSION TITLE
Erase object during loop

### DISCUSSION THREAD

**Discussion by Xenov-X** [nlohmann/json#3750] [Q&A] [answered]
I'm using json for a basic queuing function in a single threaded program. I've included a minimal working block below to hopefully explain better, but essentially when iterating through the task objects, I'd like to remove each object from the source JSON (my_json). (Normally the removal is subject to a variety of conditions, but it doesn't apply in the example. Can anyone suggest what I might be doing wrong? (included some of the things I tried as comments) 

Also worth noting the taskID will always be unique, which I guess could present an alternate method of identifying the objects if the loop iterator keys aren't useable for this? 

Cheers! 

```
#include "json.hpp"
#include <iostream>

using json = nlohmann::json;

json my_json = {
			{"id",123},
			{"tasks",{
				{{"ID",1}, {"action",1}, {"taskID",1}},
				{{"ID",1}, {"action",1}, {"taskID",2}},
				{{"ID",1}, {"action",1}, {"taskID",3}},
				{{"ID",1}, {"action",1}, {"taskID",4}}
				}
			}
		};

int main() {
	std::cout << my_json << std::endl;

	for (auto it : my_json["tasks"]) //testing alternate loop set up
	{
		std::cout << "value: " << it << '\n';
		//Processing actions here 
	
		//my_json["tasks"].erase(it); //wont compile
	}

	for (auto& el : my_json["tasks"].items())
	{
		std::cout << "key: " << el.key() << ", value:" << el.value() << '\n';
		//Processing actions here 
	
		//my_json["tasks"].erase(el);									//wont compile
		//my_json["tasks"].erase(el.key());								//error on run - cannot use erase() with ", type_name()), this));
		//my_json["tasks"].erase(my_json.begin() + std::stoi(el.key()));
		//my_json.erase(el.key());										//nothing deleted, but think this is looking a level up in the json
		//el.erase();													//wont compile
	}

	std::cout << my_json << std::endl; //should be {"id":123} or {"id":123, "tasks": null}

}
```

**Comment by falbrechtskirchinger** [comment, marked as answer]
```cpp
for (auto it : my_json["tasks"])
```
`it` is not an iterator.

```cpp
for (auto& el : my_json["tasks"].items())
```
`el` is an iterator but not one usable with `erase()`.

Use these functions to obtain valid iterators for [`erase()`](https://json.nlohmann.me/api/basic_json/erase/):
https://json.nlohmann.me/api/basic_json/begin/
https://json.nlohmann.me/api/basic_json/end/

**Reply by Xenov-X** [reply]
Thanks for the response. I've had a pay around with the begin / end functions but still don't seem to be able to crack it. Both methods of when using begin for a valid iterator get me the same "cannot seek vector after end" error. Googling seems to suggest I'm attempting to take actions out of the bounds of the vector, but I'm not following how that's the case here. 

2 test cases toggled with the "testcase" int. 
```
#include "json.hpp"
#include <iostream>

using json = nlohmann::json;

json my_json = {
			{"id",123},
			{"tasks",{
				{{"ID",1}, {"action",1}, {"taskID",1}},
				{{"ID",1}, {"action",1}, {"taskID",2}},
				{{"ID",1}, {"action",1}, {"taskID",3}},
				{{"ID",1}, {"action",1}, {"taskID",4}}
				}
			}
		};

int main() {
	
	int testcase = 0; //or 1


		if (testcase == 0) {
			for (auto it = my_json["tasks"].begin(); it != my_json["tasks"].end(); ++it)
			{
				//std::cout << *it << std::endl;
				//my_json.erase(it); //incorrect container - as expected
				my_json["tasks"].erase(it);  //cannot seek vector iterator after end

			}
		}
		else {
			//1
			for (auto& el : my_json["tasks"].items())
			{
				std::cout << "key: " << el.key() << ", value:" << el.value() << '\n';
				int key = stoi(el.key());
				std::cout << key << std::endl;
				my_json["tasks"].erase(my_json["tasks"].begin() + key); //cannot seek vector iterator after end
			}
		};

	std::cout << " Final " << my_json << std::endl; //should be {"id":123} or {"id":123, "tasks": null}

}
```

**Reply by falbrechtskirchinger** [reply]
```cpp
for (auto it = my_json["tasks"].begin(); it != my_json["tasks"].end(); ++it)
```
Think about this line for a second. You're incrementing the iterator unconditionally, including when you've just erased an element.

Here's what you should do instead:
```cpp
			for (auto it = my_json["tasks"].begin(); it != my_json["tasks"].end(); )
			{
				if(/* remove element? */)
					it = my_json["tasks"].erase(it);
				else
					++it;
			}
```

(Note that `erase()` returns an iterator. This is important, because it handles edge cases correctly.)

**Reply by Xenov-X** [reply]
Seems I need to brush up on iterators! Thank you so much for the help, your example worked a charm.
